---
title: "Analysis of cell type fractions estimated from TCGA gene expression and DNA methylation data"
author: "Wei Sun"
date: "`r Sys.Date()`"
output: 
  html_document:
    theme: journal
    highlight: tango
    toc: true
    toc_depth: 3
    toc_float:
      collapsed: true
      smooth_scroll: false
    number_sections: true
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

```{r echo = TRUE, results = 'hide', warning = FALSE, message = FALSE}

library(data.table)
library(ggplot2)
library(ggpubr)
library(ggpointdensity)
library(survival)
library(survminer)
library(readxl)
library(stargazer)
library(ggcorrplot)
library(tidyr)
theme_set(theme_bw())
```

# Loading data

Load the results of clinical information from TCGA pan cancer study.
```{r warning = FALSE, message = FALSE}
cancer_types = c("COAD", "LUAD", "LUSC", "SKCM")

cdr = read_excel("clinical_data/TCGA-CDR-SupplementalTableS1.xlsx", 
                 sheet = "TCGA-CDR")
dim(cdr)
cdr[1:2,]

cdr = cdr[which(cdr$type %in% cancer_types),]
table(cdr$type)
```

```{r include=FALSE}
# check the survival time consistency between the TCGA pan cancer study 
# and our earlier work
sc = readRDS("clinical_data/COAD_somatic_clinic.rds")
dim(sc)
names(sc)

sc = sc[,-(23:32)]
dim(sc)
sc[1:2,]

table(sc$barcode %in% cdr$bcr_patient_barcode)
cdr.COAD = cdr[match(sc$barcode, cdr$bcr_patient_barcode),]
dim(cdr.COAD)
cdr.COAD[1:2,]

plot(cdr.COAD$OS.time, sc$Time)
table(cdr.COAD$OS, sc$Delta)
```

Load the results of cell type fraction, and then merage them with clinical information. Save the information for each cancer type into a list named ```ctf``` (cell type fractions). 

```{r}
cdr$patient_id = substring(cdr$bcr_patient_barcode, first=9)
ctf = list()

for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")

  load(sprintf("deconv_expr_%s.RData", c1))
  load(sprintf("deconv_methy_%s.RData", c1))
  
  deconv_meth = rho

  cat("\ncell type fraction estimates from expression\n")
  print(dim(deconv_expr))
  print(deconv_expr[1:2,])
  
  cat("\ncell type fraction estimates from methylation\n")
  print(dim(deconv_meth))
  print(deconv_meth[1,,])
  
  stopifnot(all(rownames(deconv_expr) == dimnames(deconv_meth)[[1]]))

  ct = deconv_expr
  colnames(ct) = paste0(colnames(ct), ".E")
  ct = cbind(ct, deconv_meth[,,"EMeth"])
  
  dim(ct)
  ct[1:2,]
  ct[which(ct < 5e-3)] = 5e-3

  stopifnot(all(rownames(ct) %in% cdr$patient_id))
  ct = as.data.frame(ct)
  ct$patient_id = rownames(ct)
  ct = merge(ct, cdr, by="patient_id")
  dim(ct)
  ct[1:2,]

  ctf[[c1]] = ct
}
```

# Asess the association between cell type fractions and survival time

## Association with each cell type

This function ```cox_summary``` summarizes the association between cell type composition and survival time.

```{r warning = FALSE, message = FALSE}

cox_summary <- function(cell_type1, ct, surv_var){
  
  fm1 = sprintf("Surv(%s.time, %s) ~", surv_var, surv_var)
  fm1 = paste(fm1, "age_at_initial_pathologic_diagnosis + gender +")
  fm2 = as.formula(paste(fm1, cell_type1))
  res.cox = coxph(fm2, data = ct)
  
  x = summary(res.cox)
  k = nrow(x$coefficients)
  pval = signif(x$coefficients[k,"Pr(>|z|)"], digits=2)
  beta = signif(x$coefficients[k, "coef"], digits=3)
  HR   = signif(x$coefficients[k, "exp(coef)"], digits=3)
  HR.confint.lower = sprintf("%.2e", x$conf.int[k, "lower .95"])
  HR.confint.upper = sprintf("%.2e", x$conf.int[k, "upper .95"])
  HR   =  paste0(HR, " (", HR.confint.lower, "--", HR.confint.upper, ")")
  res = c(beta, HR, pval)
  names(res) = c("beta", "HR (95% CI for HR)", "p.value")
  return(res)
}
```

Check the associations between cell type fractions and disease-specific survival. 
```{r warning = FALSE, message = FALSE}
for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")
  
  ct = ctf[[c1]]
  cell_types = names(ct)[2:15]
  for(ctype1 in cell_types){
    ct[[ctype1]] = log2(ct[[ctype1]])
  }
  df1 = t(sapply(cell_types, cox_summary, ct=ct, surv_var="DSS"))
  print(as.data.frame(df1))
}
```


```{r warning = FALSE, message = FALSE, include=FALSE}
# Check the associations between cell type fractions and 
# progression-free survival. 

for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")

  ct = ctf[[c1]]
  cell_types = names(ct)[2:15]
  
  for(ctype1 in cell_types){
    ct[[ctype1]] = log2(ct[[ctype1]])
  }
  
  df1 = t(sapply(cell_types, cox_summary, ct=ct, surv_var="PFI"))
  print(as.data.frame(df1))
}
```



```{r warning = FALSE, message = FALSE, include=FALSE}
## Association with interactions of cell type fractions

cox_summary2 <- function(cell_types, ct, surv_var){
  cell_type1 = cell_types[1]
  cell_type2 = cell_types[2]
  fm0 = sprintf("Surv(%s.time, %s) ~", surv_var, surv_var)
  fm0 = paste(fm0, "age_at_initial_pathologic_diagnosis + gender")
  
  fm1 = as.formula(sprintf("%s + log2(%s)", fm0, cell_type1))
  fm2 = as.formula(sprintf("%s + log2(%s)*log2(%s)", fm0, 
                           cell_type1, cell_type2))
  
  res.cox1 = coxph(fm1, data = ct)
  res.cox2 = coxph(fm2, data = ct)

  x = anova(res.cox1, res.cox2)
  return(x[2,"P(>|Chi|)"])
}

for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")
  
  ct = ctf[[c1]]
  cell_types = rbind("CD8T.E", names(ct)[c(2,4:8)])
  cell_types = cbind(cell_types, rbind("CD8T", names(ct)[c(9,11:15)]))

  df1 = apply(cell_types, 2, cox_summary2, ct=ct, surv_var="DSS")
  df1 = data.frame(df1)
  rownames(df1) = apply(cell_types, 2, paste, collapse="-")
  print(df1)
}
```

## Association with relative cell type fractions

Check the associations between disease-specific survival and relative cell type fractions with respect to CD8 T cells.

```{r warning = FALSE, message = FALSE}
for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")
  
  ct = ctf[[c1]]
  cell_types = setdiff(names(ct)[2:15], c("CD8T.E", "CD8T"))
  for(ctype1 in cell_types){
    ctype1r = paste0(ctype1, "_lr")
    ctype2  = "CD8T"
    if(grepl(".E", ctype1, fixed=TRUE)){ ctype2 = "CD8T.E"}
    # cat(ctype1, ctype2, "\n")
    ct[[ctype1r]] = log2(ct[[ctype1]]/ct[[ctype2]])
  }
  
  cell_types = paste0(cell_types, "_lr")
  df1 = t(sapply(cell_types, cox_summary, ct=ct, surv_var="DSS"))
  print(as.data.frame(df1))
}
```

Generate plots of survival curves with respect to **cell type fractions** for SKCM.

```{r warning = FALSE, message = FALSE, fig.height=5, fig.width=4}

ct = ctf[["SKCM"]]
cell_types = c("CD4T", "CD8T", "Monocyte", "B", "NK", "Neutrophil", "Treg")

glist = list()
for(ctype1 in cell_types){
  ctype1E = paste0(ctype1, ".E")
  ab = rep("Low", nrow(ct))
  ab[which(ct[[ctype1E]] > median(ct[[ctype1E]]))] = "High"
  
  fit1 = survfit(Surv(DSS.time, DSS) ~ ab, data=ct)
  g1 = ggsurvplot(fit1, data = ct, risk.table = TRUE, 
                  pval=TRUE, pval.method=TRUE, 
                  title=paste(ctype1, ".E"))
  
  ab = rep("Low", nrow(ct))
  ab[which(ct[[ctype1]] > median(ct[[ctype1]]))] = "High"

  fit2 = survfit(Surv(DSS.time, DSS) ~ ab, data=ct)
  g2 = ggsurvplot(fit2, data = ct, risk.table = TRUE, 
                  pval=TRUE, pval.method=TRUE, 
                  title=ctype1)

  print(g1)
  print(g2)
}
```

Generate plots of survival curves with respect to **relative cell type fractions** for SKCM.

```{r warning = FALSE, message = FALSE, fig.height=5, fig.width=4}

ct = ctf[["SKCM"]]
cell_types = c("CD4T", "Monocyte", "B", "NK", "Neutrophil", "Treg")

glist = list()
for(ctype1 in cell_types){
  ctype1E = paste0(ctype1, ".E")
  ct.lrE  = log2(ct[[ctype1E]]/ct[["CD8T.E"]])
  rel.ab  = rep("Low", length(ct.lrE))
  rel.ab[which(ct.lrE > median(ct.lrE))] = "High"
  
  fit1 = survfit(Surv(DSS.time, DSS) ~ rel.ab, data=ct)
  g1 = ggsurvplot(fit1, data = ct, risk.table = TRUE, 
                  pval=TRUE, pval.method=TRUE, 
                  title=paste(ctype1, ".E"))
  
  ct.lr = log2(ct[[ctype1]]/ct[["CD8T"]])
  rel.ab = rep("Low", length(ct.lr))
  rel.ab[which(ct.lr > median(ct.lr))] = "High"

  fit2 = survfit(Surv(DSS.time, DSS) ~ rel.ab, data=ct)
  g2 = ggsurvplot(fit2, data = ct, risk.table = TRUE, 
                  pval=TRUE, pval.method=TRUE, 
                  title=ctype1)

  print(g1)
  print(g2)
}
```

# Assess the associations between mutation load and cell type fractions

First load mutation burden data

```{r warning = FALSE, message = FALSE}
scL = list()
for(c1 in cancer_types){
  sc = readRDS(sprintf("clinical_data/%s_somatic_clinic.rds", c1))
  dim(sc)
  names(sc)
  
  sc = sc[,-(23:32)]
  dim(sc)
  sc[1:2,]
  
  names(sc)[1] = "bcr_patient_barcode"
  ct = ctf[[c1]]
  dim(ct)
  ct[1:2,]

  sc = merge(sc, ct, by="bcr_patient_barcode")
  dim(sc)
  sc[1:2,]

  scL[[c1]] = sc
}

lapply(ctf, dim)
lapply(scL, dim)
```

Next assess the associations between mutation load and cell type fractions 

```{r warning = FALSE, message = FALSE, fig.height=4, fig.width=6}
for(c1 in cancer_types){
  cat("\n--------------------------------------------------------------\n")
  cat(c1)
  cat("\n--------------------------------------------------------------\n")

  sc = scL[[c1]]
  mb1 = sc$raw_MB_hg38_SNV + sc$raw_MB_hg38_INDEL
  mb2 = sc$raw_MB_hg38_SNV*sc$prop_clonal + sc$raw_MB_hg38_INDEL
  mb  = cbind(mb1, mb2)
  mb[mb < 10] = NA
  mb = log10(mb)
  cf  = sc[,30:43]
  
  cr1 = cor(mb, cf, method="spearman", use="pairwise.complete.obs")
  pv1 = matrix(NA, nrow=2, ncol=14)
  
  for(j in 1:2){
    for(k in 1:14){
      cjk = cor.test(mb[,j], cf[,k], method="spearman")
      pv1[j,k] = cjk$p.value
    }
  }

  cp = cbind(round(t(cr1),3), signif(t(pv1),2))
  colnames(cp)[3:4] = c("mb1.pval", "mb2.pval")
  print(cp)

  crE = cr1[,1:7]
  colnames(crE) = gsub(".E", "", colnames(crE))
  gc1 = ggcorrplot(crE, lab = TRUE, title=paste0(c1, "-Expr"))
  
  gc2 = ggcorrplot(cr1[,8:14], lab = TRUE, title=paste0(c1, "-Meth"))
  
  gga = ggarrange(gc1, gc2 + rremove("legend"), ncol = 2, nrow = 1)
  print(gga)
}
```

Illustrate cell type fraction estimates
```{r warning = FALSE, message = FALSE, fig.height=4, fig.width=7}
lapply(scL, dim)
scDf = NULL
for(c1 in cancer_types){
  scDf = rbind(scDf, scL[[c1]])
}
dim(scDf)
scDf[1:2,]

names(scDf)
names(scDf)[14] = "cancer_type"

scE = scDf[,c(14,30:36)]
rhos = scE %>% pivot_longer(-cancer_type, names_to=c("cell_type"))

ggplot(rhos, aes(x=cell_type, y=value, fill=cancer_type)) +
  geom_boxplot() + labs(title="Cell type fraction by gene expression")

scM = scDf[,c(14,37:43)]
rhos = scM %>% pivot_longer(-cancer_type, names_to=c("cell_type"))

ggplot(rhos, aes(x=cell_type, y=value, fill=cancer_type)) +
  geom_boxplot() + labs(title="Cell type fraction by DNA methylation")

```

Compare cell type fraction estimates between gene expression and DNA methylation
```{r warning = FALSE, message = FALSE, fig.height=3, fig.width=12}
cell_types = c("B", "CD4T", "CD8T", "Monocyte", "NK", "Neutrophil", "Treg")

for(cc1 in cell_types){
  g1 = ggplot(scDf, aes_string(x=paste0(cc1, ".E"), y=cc1, color="cancer_type")) + 
    geom_point() + geom_smooth(method=lm) + facet_grid(~ cancer_type) + 
    theme(legend.position = "none")
  print(g1)
}

```

Compare cell type fraction estimates vs. mutation load
```{r warning = FALSE, message = FALSE, fig.height=3, fig.width=12}
cell_types = c("B", "CD4T", "CD8T", "Monocyte", "NK", "Neutrophil", "Treg")
mb = scDf$raw_MB_hg38_SNV + scDf$raw_MB_hg38_INDEL
mb[which(mb < 10)] = NA
scDf$mb = log10(mb)

for(cc1 in cell_types){
  g1 = ggplot(scDf, aes_string(x="mb", y=paste0(cc1, ".E"), color="cancer_type")) + 
    geom_point() + geom_smooth(method=lm) + facet_grid(~ cancer_type) + 
    theme(legend.position = "none")
  print(g1)

  g1 = ggplot(scDf, aes_string(x="mb", y=cc1, color="cancer_type")) + 
    geom_point() + geom_smooth(method=lm) + facet_grid(~ cancer_type) + 
    theme(legend.position = "none")
  print(g1)
}

```

# Association between hypermutation status in COAD and immune cell fractions. 

First define the hypermutation subgroup with larger number of mutations. 

```{r warning = FALSE, message = FALSE, fig.height=3, fig.width=3}
coad = scL[["COAD"]]

ggplot(coad, aes(x=log10(raw_MB_hg38_SNV)))+
  geom_histogram(color="darkblue", fill="lightblue")  +
  geom_vline(xintercept = log10(400))
coad$hyper_mut = coad[,"raw_MB_hg38_SNV"] > 400
```

Next compare the cell type fractions between hypermuted and non-hypermutated groups. 
```{r warning = FALSE, message = FALSE, fig.height=5, fig.width=4.5}
ctypes = c("CD8T.E", "CD8T", "Treg.E", "Treg")
gl = list()
for(ct1 in ctypes){
  w1 = wilcox.test(coad[[ct1]]~ coad$hyper_mut)

  gl[[ct1]] = ggplot(coad, aes_string(x="hyper_mut", y=ct1, fill="hyper_mut")) + 
    theme(legend.position = "none") + geom_violin(trim=FALSE) + 
    geom_boxplot(width=0.1, fill="white") + 
    labs(title=sprintf("Wilcox p=%.1e", w1$p.value), x="Hypermutation status")
}

ggarrange(gl[[1]], gl[[2]], gl[[3]], gl[[4]], nrow = 2, ncol = 2)
```


# Session information
```{r}
gc()
sessionInfo()
```
